File: src/net/t3kt/tctrl/model/ModelNode.java
package net.t3kt.tctrl.model;

import net.t3kt.tctrl.schema.SchemaNode;

public abstract class ModelNode<S extends SchemaNode<?>> {

    final S schemaNode;

    ModelNode(S schemaNode) {
        this.schemaNode = schemaNode;
    }

    public S getSchemaNode() {
        return schemaNode;
    }

    public String getKey() {
        return schemaNode.getKey();
    }

    public String getLabel() {
        return schemaNode.getLabel();
    }

    public String getPath() {
        return schemaNode.getPath();
    }
}

File: win/TctrlCore/Model/ModelNode.cs
using Google.Protobuf;

namespace Tctrl.Core.Model {
    public abstract class ModelNode<S> : IModelNode<S>
        where S : IMessage {

        protected ModelNode(S schema) {
            Schema = schema;
        }

        public S Schema { get; }

        public abstract string Key { get; }
        public abstract string Label { get; }
        public abstract string Path { get; }

    }
}